pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageRedeliveryController.java

```diff
@@ -0,0 +1,114 @@
+ * under the License.
+ */
+package org.apache.pulsar.broker.service.persistent;
+
+import com.google.common.collect.ComparisonChain;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.apache.bookkeeper.mledger.impl.PositionImpl;
+import org.apache.bookkeeper.util.collections.ConcurrentLongLongPairHashMap;
+import org.apache.bookkeeper.util.collections.ConcurrentLongLongPairHashMap.LongPair;
+import org.apache.pulsar.common.util.collections.ConcurrentSortedLongPairSet;
+import org.apache.pulsar.common.util.collections.LongPairSet;
+
+public class MessageRedeliveryController {
+    private final LongPairSet messagesToRedeliver;
+    private final ConcurrentLongLongPairHashMap hashesToBeBlocked;
+
+    public MessageRedeliveryController(boolean allowOutOfOrderDelivery) {
+        this.messagesToRedeliver = new ConcurrentSortedLongPairSet(128, 2);
+        this.hashesToBeBlocked = allowOutOfOrderDelivery ? null : new ConcurrentLongLongPairHashMap(128, 2);
+    }
+
+    public boolean add(long ledgerId, long entryId) {
+        return messagesToRedeliver.add(ledgerId, entryId);
+    }
+
+    public boolean add(long ledgerId, long entryId, long stickyKeyHash) {
+        if (hashesToBeBlocked != null) {
+            hashesToBeBlocked.put(ledgerId, entryId, stickyKeyHash, 0);
+        }
+        return messagesToRedeliver.add(ledgerId, entryId);
+    }
+
+    public boolean remove(long ledgerId, long entryId) {
+        if (hashesToBeBlocked != null) {
+            hashesToBeBlocked.remove(ledgerId, entryId);
+        }
+        return messagesToRedeliver.remove(ledgerId, entryId);
+    }
+
+    public int removeAllUpTo(long markDeleteLedgerId, long markDeleteEntryId) {
+        if (hashesToBeBlocked != null) {
+            for (LongPair longPair : hashesToBeBlocked.keys()) {
+                long ledgerId = longPair.first;
+                long entryId = longPair.second;
+                if (ComparisonChain.start().compare(ledgerId, markDeleteLedgerId).compare(entryId, markDeleteEntryId)
+                        .result() <= 0) {
+                    hashesToBeBlocked.remove(ledgerId, entryId);
+                }
+            }
+        }
+        return messagesToRedeliver.removeIf((ledgerId, entryId) -> {
+            return ComparisonChain.start().compare(ledgerId, markDeleteLedgerId).compare(entryId, markDeleteEntryId)
+                    .result() <= 0;
+        });
+    }
+
+    public boolean isEmpty() {
+        return messagesToRedeliver.isEmpty();
+    }
+
+    public void clear() {
+        if (hashesToBeBlocked != null) {
+            hashesToBeBlocked.clear();
+        }
+        messagesToRedeliver.clear();
+    }
+
+    public String toString() {
+        return messagesToRedeliver.toString();
+    }
+
+    public boolean containsStickyKeyHashes(Set<Integer> stickyKeyHashes) {
+        if (hashesToBeBlocked != null) {
+            for (LongPair longPair : hashesToBeBlocked.values()) {
```

Comment:
Same here, `value()` will make a copy of the set.

Reply:
Fixed.